providers/gce/gce_instances.go

```diff
@@ -92,6 +93,58 @@ func (g *Cloud) ToInstanceReferences(zone string, instanceNames []string) (refs
 	return refs
 }
 
+// cmpAddressFamily compares IPv6 and IPv4 address families so that IPv6 is smaller than IPv4.
+func cmpAddressFamily(a, b string) int {
+	// IPv6 addresses contain a colon („:”), IPv4 addresses do not.
+	if strings.Contains(a, ":") && !strings.Contains(b, ":") {
+		return -1
+	} else if !strings.Contains(a, ":") && strings.Contains(b, ":") {
+		return 1
+	}
+
+	return 0
+}
+
+// cmpAddressType compares IP address types so that internal address is smaller than external.
+func cmpAddressType(a, b v1.NodeAddressType) int {
+	if a == v1.NodeInternalIP && b == v1.NodeExternalIP {
+		return -1
+	} else if a == v1.NodeExternalIP && b == v1.NodeInternalIP {
+		return 1
+	}
+
+	return 0
+}
+
+// orderAddresses orders node IP addresses:
+//   - In single-stack IPv6 clusters IPv6 addresses before the IPv4 addresses,
+//     internal before external.
+//   - In other clusters IPv4 addresses before IPv6 addresses, internal before
+//     external.
+func (g *Cloud) orderAddresses(addresses []v1.NodeAddress) []v1.NodeAddress {
```

Comment:
I expect something much simpler, something as this
```
import utilnet k8s.io/utils/net

orderAddresses(preferIPv6 bool, nodeAddresses []v1.NodeAddress)
sortedAddresses := make([]v1.NodeAddress, 0, len(addresses))
// add the IP addresses with the preferred ip family first
for _, nodeAddress := range nodeAddresses {
    ip := net.ParseIP(nodeAddress.Address)
    // non IP addresses as hostname will get ip nil
    if ip == nil || utilnet.IsIPv6(ip) == preferIPv6 {
        sortedAddresses = append(sortedAddresses, nodeAddress)
    }
}
// second loop add the addresses that are not the prefered ip family
for _, nodeAddress := range nodeAddresses {
    ip := net.ParseIP(nodeAddress.Address)
    // non IP addresses as hostname will get ip nil
    if ip != nil && utilnet.IsIPv6(ip) != preferIPv6 {
        sortedAddresses = append(sortedAddresses, nodeAddress)
    }
}
return sortedAddresses
```

Reply:
While this works too and most likely is slightly faster [O(n) vs O(nlogn)], I think this would be more complicated to comprehend. The code is longer than the sorting approach and requires parsing IP addresses (and handling errors if it fails). If you like it more, though, please let me know and I'll change to your approach.

Reply:
this is  consistent with existing kubernetes/kubernetes code that will help , you can read the logic directly, there is no indirection layer as in the `cmpAddressFamily`. This function also  is not considering hostnames, only IP addresses  ... I prefer to use my suggestion

Reply:
see https://github.com/kubernetes/kubernetes/blob/caafc211407c6601c6975e218502250c8f7669e6/staging/src/k8s.io/cloud-provider/node/helpers/address.go#L106 for reference

Reply:
Done